sympy/solvers/solvers.py

```diff
@@ -995,26 +995,86 @@ def _sympified_list(w):
             return [], set()
         return []
 
+    # Abs handling
+    abs_ = {}
     for i, fi in enumerate(f):
-        # Abs
-        while True:
-            was = fi
-            fi = fi.replace(Abs, lambda arg:
-                separatevars(Abs(arg)).rewrite(Piecewise) if arg.has(*symbols)
-                else Abs(arg))
-            if was == fi:
-                break
-
-        for e in fi.find(Abs):
-            if e.has(*symbols):
+        ai = {i: separatevars(i)
+            for i in fi.atoms(Abs) if i.has_free(*symbols)}
+        if not all(isinstance(v, Abs) for v in ai.values()):
+            fi = f[i] = fi.xreplace(ai)
+            ai = {i for i in fi.atoms(Abs) if i.has_free(*symbols)}
+        anew = set(ai) - set(abs_)
+        for a in ordered(anew):
+            ri = a.args[0].as_real_imag()
+            if any(_.has(*symbols) for _ in ri for _ in  _.atoms(re, im)):
                 raise NotImplementedError('solving %s when the argument '
-                    'is not real or imaginary.' % e)
+                'is not real or imaginary.' % a)
+            if 0 not in ri:
+                # if real, ri will be (a.args[0], 0)
+                # if imaginary, ri will be (0,I*a.args[0])
+                raise NotImplementedError('solving %s with real and '
+                'imag parts.' % a)
+            abs_[a] = sum(ri)
+
+    if abs_:
```

Comment:
This block duplicates a lot of the messy logic from the code below. I think that there should be a better way to structure this.

Reply:
It's unclear what block below you are referring to, but I made a change.